<a href="https://colab.research.google.com/github/Jency07/machine-learning/blob/main/Logistic_Regression_Vs_Naive_Bayes_Classifier_20MAI0026.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[Github link](https://github.com/Jency07/machine-learning/blob/main/Logistic_Regression_Vs_Naive_Bayes_Classifier_20MAI0026.ipynb)

**LOGISTIC REGRESSION**

---



***1.Import libraries***

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plot


***2.Mount filesystem***

In [ ]:
from google.colab import files
uploaded=files.upload()


Saving titanic_data.csv to titanic_data (2).csv


***3.Read the data file***

In [ ]:
data = pd.read_csv("titanic_data.csv")
data.head()


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

***4. Selecting the features for prediction***

In [ ]:
df=data[['Survived','Pclass','Sex','Age','Fare']]
df

Survived  Pclass     Sex   Age     Fare
0           0       3    male  22.0   7.2500
1           1       1  female  38.0  71.2833
2           1       3  female  26.0   7.9250
3           1       1  female  35.0  53.1000
4           0       3    male  35.0   8.0500
..        ...     ...     ...   ...      ...
886         0       2    male  27.0  13.0000
887         1       1  female  19.0  30.0000
888         0       3  female   NaN  23.4500
889         1       1    male  26.0  30.0000
890         0       3    male  32.0   7.7500

[891 rows x 5 columns]

***5.Feature Engineering***

In [ ]:
#Change male to 1 and femmal to 0
df['Sex']=df['Sex'].apply(lambda sex:1 if sex=='male' else 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
df

Survived  Pclass  Sex   Age     Fare
0           0       3    1  22.0   7.2500
1           1       1    0  38.0  71.2833
2           1       3    0  26.0   7.9250
3           1       1    0  35.0  53.1000
4           0       3    1  35.0   8.0500
..        ...     ...  ...   ...      ...
886         0       2    1  27.0  13.0000
887         1       1    0  19.0  30.0000
888         0       3    0   NaN  23.4500
889         1       1    1  26.0  30.0000
890         0       3    1  32.0   7.7500

[891 rows x 5 columns]

In [ ]:
df['Age'].isnull().sum()


177

In [ ]:
#Fixing missing values with median values
df['Age']=df['Age'].fillna(df["Age"].median())
df

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Survived  Pclass  Sex   Age     Fare
0           0       3    1  22.0   7.2500
1           1       1    0  38.0  71.2833
2           1       3    0  26.0   7.9250
3           1       1    0  35.0  53.1000
4           0       3    1  35.0   8.0500
..        ...     ...  ...   ...      ...
886         0       2    1  27.0  13.0000
887         1       1    0  19.0  30.0000
888         0       3    0  28.0  23.4500
889         1       1    1  26.0  30.0000
890         0       3    1  32.0   7.7500

[891 rows x 5 columns]

In [ ]:
df['Age'].isnull().sum()

0

***6. Fixing Predictors and response***

In [ ]:
X=df.drop("Survived",axis=1)
Y=df["Survived"]

***7.Split train and test dataset***

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state=25)

***8.Train the model***

In [ ]:
#Training using logistic regression
from sklearn.linear_model import LogisticRegression
logit=LogisticRegression()
logit.fit(X_train,Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

***9.Testing the model***

In [ ]:
#Finding y hat
Y_pred=logit.predict(X_test)

***10.Evaluating the model***

In [ ]:
#Confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix=confusion_matrix(Y_test,Y_pred)
confusion_matrix

array([[136,  29],
       [ 31,  72]])

In [ ]:
#Accuracy score
from sklearn.metrics import accuracy_score
accuracy_score=accuracy_score(Y_test,Y_pred)
accuracy_score

0.7761194029850746

In [ ]:
#Classification report
from sklearn.metrics import classification_report
classification_report=classification_report(Y_test,Y_pred)
print(classification_report)

              precision    recall  f1-score   support

           0       0.81      0.82      0.82       165
           1       0.71      0.70      0.71       103

    accuracy                           0.78       268
   macro avg       0.76      0.76      0.76       268
weighted avg       0.78      0.78      0.78       268



**NAIVE BAYES CLASSIFIER**


---



***1.Import libraries***

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plot


***2.Mount filesystem***

In [ ]:
from google.colab import files
uploaded=files.upload()


Saving titanic_data.csv to titanic_data (1).csv


***3.Read the data file***

In [ ]:
data1 = pd.read_csv("titanic_data.csv")
data1.head()


PassengerId  Survived  Pclass  ...     Fare Cabin  Embarked
0            1         0       3  ...   7.2500   NaN         S
1            2         1       1  ...  71.2833   C85         C
2            3         1       3  ...   7.9250   NaN         S
3            4         1       1  ...  53.1000  C123         S
4            5         0       3  ...   8.0500   NaN         S

[5 rows x 12 columns]

***4. Selecting the features for prediction***

In [ ]:
df1=data1[['Survived','Pclass','Sex','Age','Fare']]
df1

Survived  Pclass     Sex   Age     Fare
0           0       3    male  22.0   7.2500
1           1       1  female  38.0  71.2833
2           1       3  female  26.0   7.9250
3           1       1  female  35.0  53.1000
4           0       3    male  35.0   8.0500
..        ...     ...     ...   ...      ...
886         0       2    male  27.0  13.0000
887         1       1  female  19.0  30.0000
888         0       3  female   NaN  23.4500
889         1       1    male  26.0  30.0000
890         0       3    male  32.0   7.7500

[891 rows x 5 columns]

***5.Feature Engineering***

In [ ]:
#Change male to 1 and femmal to 0
df1['Sex']=df1['Sex'].apply(lambda sex:1 if sex=='male' else 0)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [ ]:
df1

Survived  Pclass  Sex   Age     Fare
0           0       3    1  22.0   7.2500
1           1       1    0  38.0  71.2833
2           1       3    0  26.0   7.9250
3           1       1    0  35.0  53.1000
4           0       3    1  35.0   8.0500
..        ...     ...  ...   ...      ...
886         0       2    1  27.0  13.0000
887         1       1    0  19.0  30.0000
888         0       3    0   NaN  23.4500
889         1       1    1  26.0  30.0000
890         0       3    1  32.0   7.7500

[891 rows x 5 columns]

In [ ]:
df1['Age'].isnull().sum()


177

In [ ]:
#Fixing missing values with median values
df1['Age']=df1['Age'].fillna(df1["Age"].median())
df1

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Survived  Pclass  Sex   Age     Fare
0           0       3    1  22.0   7.2500
1           1       1    0  38.0  71.2833
2           1       3    0  26.0   7.9250
3           1       1    0  35.0  53.1000
4           0       3    1  35.0   8.0500
..        ...     ...  ...   ...      ...
886         0       2    1  27.0  13.0000
887         1       1    0  19.0  30.0000
888         0       3    0  28.0  23.4500
889         1       1    1  26.0  30.0000
890         0       3    1  32.0   7.7500

[891 rows x 5 columns]

In [ ]:
df1['Age'].isnull().sum()

0

***6. Fixing Predictors and response***

In [ ]:
X1=df1.drop("Survived",axis=1)
Y1=df1["Survived"]

***7.Split train and test dataset***

In [ ]:
from sklearn.model_selection import train_test_split
X_train1,X_test1,Y_train1,Y_test1=train_test_split(X1,Y1,test_size=0.3,random_state=25)

***8.Train the model***

In [ ]:
#Training using naive bayes classifier
from sklearn.naive_bayes import GaussianNB
gaussian=GaussianNB()
gaussian.fit(X_train1,Y_train1)

GaussianNB(priors=None, var_smoothing=1e-09)

***9.Testing the model***

In [ ]:
#Finding y hat
Y_pred1=gaussian.predict(X_test1)

***10.Evaluating the model***

In [ ]:
#Confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix=confusion_matrix(Y_test1,Y_pred1)
confusion_matrix

array([[130,  35],
       [ 27,  76]])

In [ ]:
#Accuracy score
from sklearn.metrics import accuracy_score
accuracy_score=accuracy_score(Y_test1,Y_pred1)
accuracy_score

0.7686567164179104

In [ ]:
#Classification report
from sklearn.metrics import classification_report
classification_report=classification_report(Y_test1,Y_pred1)
print(classification_report)

              precision    recall  f1-score   support

           0       0.83      0.79      0.81       165
           1       0.68      0.74      0.71       103

    accuracy                           0.77       268
   macro avg       0.76      0.76      0.76       268
weighted avg       0.77      0.77      0.77       268





---



**COMPARISON OF LOGISTIC REGRESSION AND NAIVE BAYES CLASSIFIER**

Based on observation from above models, naive bayes gives an accuracy of 76% while 77% accuracy is achieved using logistic regression.

Below are the observations on analysis of both the models from various online sources

* Both Naive Bayes and Logistic regression are linear classifiers but logistic regression makes a prediction for the probability using a direct functional form where as naive nayes figures out how the data was generated given the results.
* Naive Bayes assumes that the features are conditionally independent.Real data sets are never perfectly independent but they can be close.
Naive Bayes has a higher bias but lower variance compared to logistic regression.
* Thus if the data set follows the bias then Naive Bayes will be a better classifier. 



---

